SourceBuffer will always perform a sync call to GPU process when there's a need to evict content.

jyavenard opened the pull request:
#### 082067e37acd774037dc66d5f3e2dd4cfe66a3f0
<pre>
SourceBuffer will always perform a sync call to GPU process when there&apos;s a need to evict content.
<a href="https://bugs.webkit.org/show_bug.cgi?id=276999">https://bugs.webkit.org/show_bug.cgi?id=276999</a>
<a href="https://rdar.apple.com/132402500">rdar://132402500</a>

Reviewed by Youenn Fablet.

Prior 275380@main, eviction was performed synchronously and the eviction data algorithm would evict 3s of data at a time
until there was sufficient buffer space available for the appendBuffer operation to complete.
In 275380@main, we pre-calculated how much data could be removed in total. However, the calculation incorrectly
looked at the initial 3s of removable data only. Which would make the SourceBuffer always consider that no data was evictable
preventing an asynchronous eviction to occur and instead generating a sync call to SourceBufferPrivate::evictData.

Added tests.

* LayoutTests/media/media-source/media-managedmse-seek-and-evictable-expected.txt: Added.
* LayoutTests/media/media-source/media-managedmse-seek-and-evictable.html: Added.
* Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
(WebCore::SourceBuffer::evictableSize const):
* Source/WebCore/Modules/mediasource/SourceBuffer.h:
* Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
(WebCore::SourceBufferPrivate::seekToTime):
(WebCore::SourceBufferPrivate::computeEvictionData):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::evictableSize):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Canonical link: <a href="https://commits.webkit.org/281326@main">https://commits.webkit.org/281326@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a23f2e47b42c95761b88adbff6d1e058f64d3d37

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/59541 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/38886 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/12065 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/63456 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/10060 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/61670 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/46539 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/10216 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/48318 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7048 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/61571 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/36306 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/51540 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29151 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/33010 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/8793 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/8988 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/54957 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/9071 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/65188 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/3469 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/8963 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/55661 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/3480 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/51533 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/55781 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13193 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/2879 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/34700 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/35783 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/36869 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/35528 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
```diff
@@ -459,12 +461,14 @@ void SourceBufferPrivate::computeEvictionData(ComputeEvictionDataRule rule)
 
         auto rangeStartBeforeCurrentTime = minimumBufferedTime();
-        auto rangeEndBeforeCurrentTime = std::min(rangeStartBeforeCurrentTime, maximumRangeEndBeforeCurrentTime);
+        auto rangeEndBeforeCurrentTime = maximumRangeEndBeforeCurrentTime;
```
Why have both rangeEndBeforeCurrentTime and maximumRangeEndBeforeCurrentTime?
Can we just have maximumRangeEndBeforeCurrentTime?

jyavenard replied:
I had tried to stay as close to the code actually performing the eviction.

eric-carlson commented on LayoutTests/media/media-source/media-managedmse-seek-and-evictable.html:
```diff
@@ -0,0 +1,127 @@
+<!DOCTYPE html>
+<html>
+<head>
+    <title>mock-media-source</title>
```
The title could be more descriptive

jyavenard committed:
SourceBuffer will always perform a sync call to GPU process when there's a need to evict content.
https://bugs.webkit.org/show_bug.cgi?id=276999
rdar://132402500

Reviewed by Youenn Fablet.

Prior 275380@main, eviction was performed synchronously and the eviction data algorithm would evict 3s of data at a time
until there was sufficient buffer space available for the appendBuffer operation to complete.
In 275380@main, we pre-calculated how much data could be removed in total. However, the calculation incorrectly
looked at the initial 3s of removable data only. Which would make the SourceBuffer always consider that no data was evictable
preventing an asynchronous eviction to occur and instead generating a sync call to SourceBufferPrivate::evictData.

Added tests.

* LayoutTests/media/media-source/media-managedmse-seek-and-evictable-expected.txt: Added.
* LayoutTests/media/media-source/media-managedmse-seek-and-evictable.html: Added.
* Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
(WebCore::SourceBuffer::evictableSize const):
* Source/WebCore/Modules/mediasource/SourceBuffer.h:
* Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
(WebCore::SourceBufferPrivate::seekToTime):
(WebCore::SourceBufferPrivate::computeEvictionData):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::evictableSize):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Canonical link: https://commits.webkit.org/281326@main
